cpp/ql/src/experimental/Security/CWE/CWE-570/WrongInDetectingAndHandlingMemoryAllocationErrors.ql

```diff
@@ -11,76 +11,180 @@
+    convertedExprMayThrow([binOp.getLeftOperand(), binOp.getRightOperand()])
+  )
+  or
+  exists(CommaExpr comma | comma = e |
+    convertedExprMayThrow([comma.getLeftOperand(), comma.getRightOperand()])
+  )
+  or
+  exists(StmtExpr stmtExpr | stmtExpr = e |
+    convertedExprMayThrow(stmtExpr.getResultExpr()) or
+    stmtMayThrow(stmtExpr.getStmt())
+  )
+  or
+  convertedExprMayThrow(e.(Conversion).getExpr())
+  or
+  exists(FunctionCall fc | fc = e |
+    not exists(fc.getTarget()) or
+    functionMayThrow(fc.getTarget()) or
+    convertedExprMayThrow(fc.getAnArgument())
+  )
+}
+
+/** An allocator that will not throw an exception. */
+class NoThrowAllocator extends Function {
+  NoThrowAllocator() {
+    exists(NewOrNewArrayExpr newExpr |
+      newExpr.getAllocator() = this and
+      not functionMayThrow(this)
     )
   }
+}
 
-  /**
-   * Holds if handler `try ... catch` exists.
-   */
-  predicate isExistsTryCatchBlock() {
-    exists(TryStmt ts | this.getEnclosingStmt() = ts.getStmt().getAChild*())
-  }
+/** An allocator that might throw an exception. */
+class ThrowingAllocator extends Function {
+  ThrowingAllocator() { not this instanceof NoThrowAllocator }
+}
 
-  /**
-   * Holds if results call `operator new` check in `operator if`.
-   */
-  predicate isExistsIfCondition() {
-    exists(IfCompareWithZero ifc, AssignExpr aex, Initializer it |
-      // call `operator new` directly from the condition of `operator if`.
-      this = ifc.getCondition().getAChild*()
-      or
-      // check results call `operator new` with variable appropriation
-      postDominates(ifc, this) and
-      aex.getAChild() = exp and
-      ifc.getCondition().getAChild().(VariableAccess).getTarget() =
-        aex.getLValue().(VariableAccess).getTarget()
-      or
-      // check results call `operator new` with declaration variable
-      postDominates(ifc, this) and
-      exp = it.getExpr() and
-      it.getDeclaration() = ifc.getCondition().getAChild().(VariableAccess).getTarget()
-    )
+/** The `std::bad_alloc` exception and its `bsl` variant. */
+class BadAllocType extends Class {
+  BadAllocType() { this.hasGlobalOrStdOrBslName("bad_alloc") }
+}
+
+/**
+ * A catch block that catches a `std::bad_alloc` (or any of its subclasses), or a catch
+ * block that catches every exception (i.e., `catch(...)`).
+ */
+class BadAllocCatchBlock extends CatchBlock {
+  BadAllocCatchBlock() {
+    this.getParameter().getUnspecifiedType() = any(BadAllocType badAlloc).getADerivedClass*()
+    or
+    not exists(this.getParameter())
   }
+}
+
+/**
+ * Holds if `newExpr` will not throw an exception, but is embedded in a `try` statement
+ * with a catch block `catchBlock` that catches an `std::bad_alloc` exception.
+ */
+predicate noThrowInTryBlock(NewOrNewArrayExpr newExpr, BadAllocCatchBlock catchBlock) {
+  exists(TryStmt try |
+    forall(Expr cand | cand.getEnclosingBlock().getEnclosingBlock*() = try.getStmt() |
+      not convertedExprMayThrow(cand)
+    ) and
```

Comment:
Can this be replaced with `not stmtMayThrow(try.getStmt())`?

Reply:
I'm not sure I understand. What does `this` refer to in your sentence?

Reply:
The entire `forall`.  If every `Expr cand` nested in the `try.getStmt()` can't throw, then the `try.getStmt()` itself can't throw.

Reply:
In the other direction it's more complicated actually - if the `try.getStmt()` can't throw, then no `Expr cand` nested in the `try.getStmt()` can throw, *except* if there's another `try` block in between that catches the exception.  I'm not sure which behaviour is best here.

Reply:
I think that exceptional case is covered by `stmtMayThrow` not traversing into the statement of `TryStmt`s, isn't it?

Reply:
Yes, but it isn't covered in your original `forall` code, so it's a change in behaviour.  Probably for the better but I'm not sure what real world examples would look like.

Reply:
Good point. I've made the change in 856d512aa677a8f3097320918d4f21d7da003128 since it (a) is aesthetically more pleasing, and (b) it deletes more code than it adds.